server/src/main/java/org/opensearch/index/IndexModule.java

```diff
@@ -518,6 +519,15 @@ public IndexService newIndexService(
             if (IndexService.needsMapperService(indexSettings, indexCreationContext)) {
                 indexAnalyzers = analysisRegistry.build(indexSettings);
             }
+            final List<String> hybridMMapExtensions = new ArrayList<>();
```

Comment:
@martin-gaievski I am really curious why we need this additional bag of providers or/and settings vs using the `IndexSettings` directly? 

Fe, the idea roughly: 

```
"settings": {
    "index.store.type": "hybridfs",
    "index.store.hybridfs.extensions": "...."
  }
```

Reply:
@reta index setting approach requires manual work for existing indexes, and the way it's implemented is override vs append. Another dimension is future additions, my view is that it's easy to allow plugins maintaining list of extensions based on a plugin's need, and that doesn't always mean a single setting of file extensions. Say list for hybridfs, it has some tradeoffs and will not work for everyone. In kNN we want to provide two different modes for user to choose from, and that list setting is only part of one mode. With this PR we can encapsulate hybridfs setting as it's internal detail of implementation and also minimize the chance of configuration error.

Reply:
The concept of index store is generic in OpenSearch and I think it should stay generic: the extensions or any additional settings should be able to be configured for any index store (including `hybridfs` for example).

If `IndexSettings` are not sufficient, it probably make sense to extend the `IndexStorePlugin` with additional factory methods / extensions. In this way the `KNN` (or any other plugin) could contribution additional modes.

Reply:
I like the idea of implementing `IndexStorePlugin` on kNN side, it makes more sense and using existing mechanism for extension. I have one doubt related to the type, mainly it's due to [this check](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L636) on core side. Seems we cannot override factory for existing storage type, only add new ones. 

For the new type how should we register and enable it from plugin? And the second question - should we add existing core extension list from `hydridfs` to our type or both storage type exist independently?

Reply:
> I have one doubt related to the type, mainly it's due to [this check](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L636) on core side. Seems we cannot override factory for existing storage type, only add new ones.

I think this is fine, we could introduce the concept of customizer or similar to apply to built-in factories 

> For the new type how should we register and enable it from plugin?

Yes

> should we add existing core extension list from hydridfs to our type or both storage type exist independently?

This part I not sure, you mentioned that KNN needs to add `hydridfs` extensions which may not be suitable to everyone, probably we could limit to KNN only

Reply:
> > I have one doubt related to the type, mainly it's due to [this check](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L636) on core side. Seems we cannot override factory for existing storage type, only add new ones.
> 
> I think this is fine, we could introduce the concept of customizer or similar to apply to built-in factories

If this case be provided I think it resolves the rest of concerns with a new store type. However I do have some doubts we should use mechanism of `IndexStorePlugin`. At the end directory factory return one Directory object, so any type of customizer on the store type level should override for core type, and I'm not sure if we want to allow it. Something on a lower level will look similar to what is currently in PR, with idea - core will apply their logic in factories and return proper Directory (hybridfs or niofs or other), but customizer (from plugin side) just provides list of additional extensions. 

> 
> > For the new type how should we register and enable it from plugin?
> 
> Yes
> 
> > should we add existing core extension list from hydridfs to our type or both storage type exist independently?
> 
> This part I not sure, you mentioned that KNN needs to add `hydridfs` extensions which may not be suitable to everyone, probably we could limit to KNN only

After checking code I think different types exist independently, we can have only "new" extensions in case of new type. In case of same type with customization it's up to implementation of the customization mechanism, say with current PR it's append only, no overrides

Reply:
> with idea - core will apply their logic in factories and return proper Directory (hybridfs or niofs or other), but customizer (from plugin side) just provides list of additional extensions.

:+1: Sounds good to me

>  In case of same type with customization it's up to implementation of the customization mechanism, say with current PR it's append only, no overrides

:+1: